File: sdk/Managed/src/Microsoft.WindowsAzure.MobileServices/Table/Query/OData/ConstantNode.cs
// ----------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
// ----------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Microsoft.WindowsAzure.MobileServices.Query
{
    /// <summary>
    /// Node representing a primitive constant value.
    /// </summary>  
    public sealed class ConstantNode: QueryNode
    {
        /// <summary>
        /// Gets the kind of the query node.
        /// </summary>
        public override QueryNodeKind Kind
        {
            get { return QueryNodeKind.Constant; }
        }

        /// <summary>
        /// Gets the primitive constant value.
        /// </summary>
        public object Value { get; private set; }

        /// <summary>
        /// Initializes an instance of <see cref="ConstantNode"/>
        /// </summary>
        /// <param name="value">This node's primitive value.</param>
        public ConstantNode(object value)
        {
            this.Value = value;
        }

        /// <summary>
        /// Accept a <see cref="QueryNodeVisitor{T}" /> to walk a tree of <see cref="QueryNode" />s.
        /// </summary>
        /// <typeparam name="T">Type that the visitor will return after visiting this token.</typeparam>
        /// <param name="visitor">An implementation of the visitor interface.</param>
        /// <returns>An object whose type is determined by the type parameter of the visitor.</returns>
        public override T Accept<T>(QueryNodeVisitor<T> visitor)
        {
            return visitor.Visit(this);
        }
    }
}

File: sdk/android/src/sdk/src/main/java/com/microsoft/windowsazure/mobileservices/table/query/ConstantNode.java
/*
Copyright (c) Microsoft Open Technologies, Inc.
All Rights Reserved
Apache 2.0 License
 
   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at
 
     http://www.apache.org/licenses/LICENSE-2.0
 
   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 
See the Apache Version 2.0 License for specific language governing permissions and limitations under the License.
 */

/**
 * ConstantNode.java
 */
package com.microsoft.windowsazure.mobileservices.table.query;

import java.util.Date;

/**
 * Class that represents a constant query node
 */
class ConstantNode implements QueryNode {
    private Object mValue;

    @Override
    public QueryNode deepClone() {
        ConstantNode clone = new ConstantNode();

        if (this.mValue instanceof Date) {
            clone.mValue = new Date(((Date) this.mValue).getTime());
        } else {
            clone.mValue = this.mValue;
        }

        return clone;
    }

    @Override
    public QueryNodeKind getKind() {
        return QueryNodeKind.Constant;
    }

    @Override
    public <T> T accept(QueryNodeVisitor<T> visitor) {
        return visitor.visit(this);
    }

    /**
     * Gets the constant node value.
     */
    Object getValue() {
        return this.mValue;
    }

    /**
     * Sets the constant node value.
     */
    void setValue(Object value) {
        this.mValue = value;
    }
}